editor/src/components/canvas/controls/new-canvas-controls.tsx

```diff
@@ -204,12 +207,32 @@ export const selectElementsThatRespectLayout = createSelector(
   },
 )
 
-const NewCanvasControlsInner = (props: NewCanvasControlsInnerProps) => {
-  const [localSelectedViews, setLocalSelectedViews] = usePropControlledStateV2(
-    props.derived.canvas.transientState.selectedViews,
+/**
+ * these local states are deliberately stored as atoms!
+ * this makes it easier to create a shortcut so we can update the selected views from ui-jsx-canvas
+ * this ONLY works with a single canvas!
+ * overhaul this for multi-canvas support
+ */
+const localSelectedViewsAtom = atom<Array<TemplatePath>>({ key: 'localSelectedViews', default: [] })
```

Comment:
We have a perceived performance trick: if you click to select something on the canvas, we only redraw the canvas controls on the first frame, _then_ dispatch the selection action to the editor. This is so that we can visually react to the user interaction the next possible frame without having to worry about the editor update's speed. This hack vastly improves the percieved performance of the editor.

The problem I need to solve now is that select-mode-hooks are used from the UiJsxCanvas component, but I need to change the state of NewCanvasControls when the selection changes. And I don't want to re-render the UiJsxCanvas, because that would jeopardize the reaction time. So I needed a way to be able to update the canvas controls without having to rerender the canvas, or having to rerender a common ancestor and dealing with memoization and/or refs to achieve the same effect.

You can think of a Recoil Atom as a "global variable". If you have reference to the _same_ Atom, it will hold the same value for all the places that are accessing it. If any one place updates the value of the atom, every subscriber (places that are calling the `useRecoilState` hook) will be re-rendered.
Recoil also supports write-only access to the Atom, which means you can have some components which can update the value of the atom, but do not need to rerender if the atom value changes. This feature made me pick Recoil over Jotai, which does not support write-only access.